gateway-sp-comms/src/shared_socket.rs

```diff
@@ -0,0 +1,712 @@
+        let old_scope_id = self.discovery_addr.scope_id();
+        let new_scope_id =
+            self.scope_id_cache.refresh_by_name(&self.interface).await?;
+
+        if new_scope_id == old_scope_id {
+            return Ok(false);
+        }
+
+        // Scope ID changed; if we had a nonzero scope ID, we'd previously
+        // joined the SP_TO_MGS multicast group on it; now leave it.
+        if old_scope_id != 0 {
+            if let Err(err) = self
+                .socket
+                .leave_multicast_v6(&SP_TO_MGS_MULTICAST_ADDR, old_scope_id)
+            {
+                // This presumably isn't fatal, because `old_scope_id` almost
+                // certainly references an interface that no longer exists; just
+                // log a warning and move on.
+                warn!(
+                    self.log, "failed to leave multicast group";
+                    "group" => %SP_TO_MGS_MULTICAST_ADDR,
+                    "interface" => self.interface(),
+                    "scope_id" => old_scope_id,
+                    "err" => %err,
+                );
+            }
+        }
+
+        // Join the same multicast group on our new scope ID.
+        self.socket
+            .join_multicast_v6(&SP_TO_MGS_MULTICAST_ADDR, new_scope_id)
+            .map_err(|err| SingleSpHandleError::JoinMulticast {
+                group: SP_TO_MGS_MULTICAST_ADDR,
+                interface: self.interface.to_string(),
+                err,
+            })?;
+
+        self.discovery_addr.set_scope_id(new_scope_id);
+        debug!(
+            self.log, "refreshed scope ID for SP interface";
+            "interface" => self.interface(),
+            "discovery_addr" => %self.discovery_addr(),
+        );
+        Ok(true)
+    }
+
+    pub(crate) async fn send(
+        &mut self,
+        data: &[u8],
+    ) -> Result<(), SingleSpHandleError> {
+        // Is this the first time we're being used? Update our scope ID.
+        if self.discovery_addr.scope_id() == 0 {
+            self.refresh_scope_id().await?;
+        }
+
+        loop {
+            let err = match self.socket.send_to(data, self.discovery_addr).await
+            {
+                Ok(n) => {
+                    // We should never be asked to send more data than will fit
+                    // in one send; assert we didn't send an incomplete packet.
+                    assert_eq!(n, data.len(), "UDP send_to incomplete");
+                    return Ok(());
+                }
+                Err(err) => err,
+            };
+
+            // UDP send failures are relatively rare; a likely case is that the
+            // interface we're using has been deleted (and hopefully
+            // recreated!). If we fail to send, we will try evicting our cached
+            // scope ID to see if it's changed; if it hasn't, return the send
+            // error.
+            if self.refresh_scope_id().await? {
```

Comment:
I don't know how this could possibly happen, but it would be weird to have the scope id continuously refresh and trigger an infinite loop. We could provide a max count, but that may be overkill.

Reply:
I don't think it hurts to put a max count, good idea. In practice I think the only way we could hit that is if something _else_ is in a (fast!) retry loop destroying and recreating interfaces, but if that's happening it seems better to log a "the scope ID has changed 5 times, we gave up" error than spin forever.

Reply:
Fixed in f7b545598e20dca33228f09c77812caccc9f3d32